src/na/nasp/SolverFactory.cpp

```diff
@@ -0,0 +1,114 @@
+#include "na/nasp/SolverFactory.hpp"
+
+#include "ir/QuantumComputation.hpp"
+#include "ir/operations/OpType.hpp"
+#include "na/Architecture.hpp"
+#include "na/NADefinitions.hpp"
+#include "na/nasp/Solver.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <cstdint>
+#include <sstream>
+#include <stdexcept>
+#include <utility>
+#include <vector>
+
+namespace na {
+auto SolverFactory::create(const Architecture& arch) -> NASolver {
+  const auto interactionZone =
+      *arch.getPropertiesOfOperation({qc::Z, 1}).zones.cbegin();
+  const auto maxX =
+      static_cast<std::uint16_t>(arch.getNColsInZone(interactionZone) - 1);
+  const auto maxEntanglingY =
+      static_cast<std::uint16_t>(arch.getNrowsInZone(interactionZone) - 1);
+  const auto maxC =
+      static_cast<std::uint16_t>(arch.getPropertiesOfShuttlingUnit(0).cols);
+  const auto maxR =
+      static_cast<std::uint16_t>(arch.getPropertiesOfShuttlingUnit(0).rows);
+  const auto storageZone = arch.getInitialZones().front();
+  const auto maxY = static_cast<std::uint16_t>(
+      maxEntanglingY + arch.getNrowsInZone(storageZone));
```

Comment:
Quite a few casts here that might be possibly avoided by changing up some of the underlying data types.

Reply:
The issue is that the architecture can and should handle larger values than the solver will ever do. So unless I use unnecessary large datatype for the solver, those casts are inavoidable.